### PR TITLE
[Test] Make test_serialization.py device-agnostic for out-of-tree backends

### PR DESCRIPTION
**Pull request by RiyaP2508**
## Summary

Refactors `test/test_serialization.py` to separate device-dependent tests from CPU-only tests, following the same pattern used in `test_dropout.py` and `test_convolution.py`.

Creates a new `TestSerializationDeviceType(TestCase)` class registered with `instantiate_device_type_tests`, enabling automatic test generation for all device backends (CPU, CUDA, XPU, OpenReg, etc.) instead of relying on manual `torch.cuda.is_available()` guards.

### Tests moved to `TestSerializationDeviceType`:

| Test | Source | Decorator |
|------|--------|-----------|
| `test_serialization_map_location` | `SerializationMixin` | `@onlyAccelerator` |
| `test_save_different_dtype_unallocated` | `SerializationMixin` | — |
| `test_save_different_dtype_error` | `SerializationMixin` | — |
| `test_skip_data_load` | `SerializationMixin` | — |
| `test_serialization_mmap_loading_with_map_location` | `TestSerialization` | `@onlyCUDA` |
| `test_skip_data_serialization` | `TestSerialization` | — |
| `test_use_pinned_memory_for_d2h` | `TestSerialization` | `@onlyAccelerator` |

### What stays unchanged:
- **`TestBothSerialization`** — already device-generic via `instantiate_device_type_tests`
- **`TestOldSerialization`** — CPU-only (legacy format); retains CPU-only versions of `test_save_different_dtype_unallocated`, `test_save_different_dtype_error`, and `test_serialization_map_location` through the mixin
- **`TestSerialization`** — all remaining ~67 tests are CPU/meta/FakeTensor only
- **`TestSubclassSerialization`** — 1 CUDA `@skipIf` left as-is (not worth a class split for a single test)

## Test plan
- [x] All 225 tests pass (191 passed, 20 skipped, 2 deselected — same as upstream)
- [x] `TestSerializationDeviceTypeCPU` and `TestSerializationDeviceTypeCUDA` variants are generated correctly
- [x] `TestOldSerialization` still inherits and passes CPU-only mixin tests
- [x] No new test regressions

cc: @fffrog @albanD

### CODE REVIEW

File: test/test_serialization.py [hunk 1]
```diff
@@ -4952,6 +4761,224 @@ def run(self, *args, **kwargs):
...
+            ft = converter.from_real_tensor(mode, torch.randn(2, device=device))
+            exc = pickle.PicklingError if sys.version_info >= (3, 14) else AttributeError
+            with self.assertRaisesRegex(
+                exc,
+                r"Can't (get|pickle) local object (<function |')WeakValueDictionary\.__init__\.<locals>\.remove"
+            ):
+                with skip_data(), BytesIOContext() as f:
+                    torch.save(ft, f)
+
+    @onlyAccelerator
```
Comment (fffrog) [inline test/test_serialization.py:5045]: We should skip MPS, right?

https://github.com/pytorch/pytorch/blob/d607ab6c28414b18fdaffdf8b890882d154d635b/test/test_serialization.py#L4722

Comment (RiyaP2508) [reply test/test_serialization.py:5045]: Good catch! Yes, need to add @skipMPS since the pin memory allocator isn't registered on MPS. Will add:

```
  @onlyAccelerator
  @skipMPS("pin memory allocator is not registered on MPS")
  def test_use_pinned_memory_for_d2h(self, device):
```

File: test/test_serialization.py [hunk 2]
```diff
@@ -4952,6 +4761,224 @@ def run(self, *args, **kwargs):
...
+                return self.fc2(self.fc1(input))
+
+        with TemporaryFileName() as f:
+            with torch.device(device):
+                m = DummyModel()
+            state_dict = m.state_dict()
+            torch.save(state_dict, f)
+            result = torch.load(f, mmap=True)
+            for v in result.values():
+                self.assertTrue(v.is_cuda)
```
Comment (fffrog) [inline test/test_serialization.py]:
```suggestion
                self.assertEqual(v.device.type, device.type)
```

**Review by fffrog — APPROVED** [review]
SGTM, thank you

**Comment by fffrog** [comment]
Just a side thread: I noticed there are some other test cases that can be refactored to be device-agnostic. They aren't included in the current PR, any plan about those? :D

**Comment by RiyaP2508** [comment]
> Just a side thread: I noticed there are some other test cases that can be refactored to be device-agnostic. They aren't included in the current PR, any plan about those? :D

Yes, there are  few other tests that could be refactored! I kept this PR focused on the tests in TestSerialization and TestOldSerialization, but there are some in TestSubclassSerialization as well. I can tackle those in a follow-up PR to keep things scoped. Happy to adjust if you would like otherwise.

File: test/test_serialization.py (hunk 1 above)
Comment (RiyaP2508) [reply test/test_serialization.py:5045]: Updated these suggestions in the latest push. Thank you for the feedbacks!

File: test/test_serialization.py [hunk 3]
```diff
@@ -4952,6 +4761,225 @@ def run(self, *args, **kwargs):
...
+            self.assertEqual(
+                sd_loaded['tt'].a.untyped_storage().nbytes(),
+                sd['tt'].a.untyped_storage().nbytes()
+            )
+            for k in sd_loaded:
+                sd_loaded[k] = sd_loaded[k].zero_()
+            self.assertEqual(sd_loaded, sd_zeroed)
+
+    @onlyCUDA
+    def test_serialization_mmap_loading_with_map_location(self, device):
```
Comment (jbschlosser) [inline test/test_serialization.py:4976]: I think this is a pre-existing issue but I'm a bit confused by this test name since I don't see the `map_location` arg used in this test :p

also should this test ideally be `onlyCUDA` or `onlyAccelerator` (with appropriate skips)?

**Comment by jbschlosser** [comment]
@claude review this and point out in detail all changes made that are not strictly code movement of tests into the new `TestSerializationDeviceType` test class

File: test/test_serialization.py [hunk 4]
```diff
@@ -649,18 +644,6 @@ def load_bytes():
             'cpu',
             torch.device('cpu'),
         ]
-        gpu_0_map_locations = generate_map_locations('cuda')
-        gpu_last_map_locations = [
```
Comment (jbschlosser) [inline test/test_serialization.py:653]: Claude pointed out this for `test_serialization_map_location`:

> Behavioral difference: The original tested multi-device map locations (device 0 and the last device) on CUDA/XPU/MTIA. The new device-type version only tests device 0 (the device provided by the test framework). This is a loss of coverage for multi-device scenarios.

I think we should maintain last-device coverage as well for new device-agnostic version of the test - wdyt?

File: test/test_serialization.py [hunk 5]
```diff
@@ -890,10 +852,6 @@ def test_load_error_msg(self):
             torch.load(resource, weights_only=False)
 
     def test_save_different_dtype_unallocated(self):
-        devices = ['cpu']
```
Comment (jbschlosser) [inline test/test_serialization.py:893]: hm should we have two versions of this test spread across two test classes? Ideally I think we could have one test that works for both cpu & other devices, but let me know if this a lot of trouble

File: test/test_serialization.py [hunk 6]
```diff
@@ -902,52 +860,47 @@ def save_load_check(a, b):
...
                 s = torch.TypedStorage(
                     wrap_storage=a.storage().untyped(),
                     dtype=other_dtype)
                 save_load_check(a, s)
                 save_load_check(a.storage(), s)
-                b = torch.tensor([], dtype=other_dtype, device=device)
+                b = torch.tensor([], dtype=other_dtype)
                 save_load_check(a, b)
 
     def test_save_different_dtype_error(self):
```
Comment (jbschlosser) [inline test/test_serialization.py]: similarly for this test, maybe we can simplify to a single version of this test in the new device test class

File: test/test_serialization.py [hunk 7]
```diff
@@ -4952,6 +4761,225 @@ def run(self, *args, **kwargs):
...
+            with torch.device(device):
+                m = DummyModel()
+            state_dict = m.state_dict()
+            torch.save(state_dict, f)
+            result = torch.load(f, mmap=True)
+            for v in result.values():
+                self.assertEqual(v.device.type, device.type)
+
+    @parametrize("materialize_fake", (True, False))
+    def test_skip_data_serialization(self, device, materialize_fake):
```
Comment (jbschlosser) [inline test/test_serialization.py:4996]: nit: Claude pointed out that 3 comments have been removed from this test (https://github.com/pytorch/pytorch/pull/184261#issuecomment-4522165498); I think it's best to keep them in if still applicable

**Review by jbschlosser — COMMENTED** [review]
Thanks for the hard work here!

File: test/test_serialization.py (hunk 3 above)
Comment (RiyaP2508) [reply test/test_serialization.py:4976]: sorry for the delayed response.
Good point on the name - it's pre-existing from upstream. The test validates that mmap loading works when tensors have non-CPU location tags, not that a map_location argument is passed. Didn't rename in this PR to keep the diff focused on the device-type extraction, but happy to rename in a follow-up.

Re @onlyCUDA vs @onlyAccelerator: kept @onlyCUDA in the latest push since the mmap + device deserialization path hasn't been validated on other accelerators - broadening without that could introduce false failures. Can revisit once other backends confirm support.

File: test/test_serialization.py (hunk 4 above)
Comment (RiyaP2508) [reply test/test_serialization.py:653]: Agreed - addressed in the latest push. Added test_serialization_map_location_multi_device with @deviceCountAtLeast(2) that tests loading to the last device (devices[-1]), restoring the multi-device coverage.

File: test/test_serialization.py (hunk 5 above)
Comment (RiyaP2508) [reply test/test_serialization.py:893]: Consolidated in the latest push - removed the mixin version and kept only the device-type version. TestSerializationDeviceTypeCPU covers the CPU path automatically via instantiate_device_type_tests.

File: test/test_serialization.py (hunk 6 above)
Comment (RiyaP2508) [reply test/test_serialization.py]: updated in latest commit

File: test/test_serialization.py (hunk 7 above)
Comment (RiyaP2508) [reply test/test_serialization.py:4996]: Restored all three comments in the latest push.

Thanks for the feedbacks Joel !!

**Comment by jansel** [comment]
@claude explain those CI failures, are they related?

**Review by jansel — CHANGES_REQUESTED** [review]
fix test failures

**Review by jansel — APPROVED** [review]
